tests/cpp/test_alias.cpp

```diff
@@ -1545,4 +1545,38 @@ TEST_F(AliasTest, Issue2664) {
       __FILE__);
 }
 
+TEST_F(AliasTest, AliasOutputNoSegmentation) {
+  // testing a complete fusion
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const std::vector<int64_t> in_shape({2, 3, 4});
+
+  TensorView* in = makeContigConcreteTensor(in_shape);
+  fusion->addInput(in);
+  TensorView* out = add(in, IrBuilder::create<Val>(3.141));
+  fusion->addOutput(out);
+  // this is an inplace update and shouldn't be segmented into its own kernel by alias analysis
+  TensorView* update_input = set(out);
+  fusion->aliasOutputToInput(update_input, in, AllocationType::ReuseBuffer);
+
+  FusionExecutorCache executor_cache(std::move(fusion));
+  at::Tensor in_tensor =
+      at::randn(in_shape, at::dtype(at::kFloat).device(at::kCUDA, 0));
+  at::Tensor original_tensor = in_tensor.clone();
+  std::vector<at::Tensor> out_tensors =
+      executor_cache.runFusionWithInputs({in_tensor});
+  ASSERT_EQ(out_tensors.size(), 1);
+
+  // Verify inplace update
+  EXPECT_TRUE(out_tensors[0].equal(in_tensor));
+  // Verify no segmentation
+  EXPECT_FALSE(executor_cache.getMostRecentKernelRuntime()->isSegmented())
```

Comment:
Question: does nvFuser support in-place updates (aka input/output alias) in multi-segment fusions? I vaguely remember seeing some segmentation code dealing with that, but I'm not sure.

Reply:
Yes we do.

@Priya2698 actually added a preseg pass to force segmentation to avoid RW racing, that's probably what you are referring to.

Reply:
I was referring to https://github.com/NVIDIA/Fuser/blob/7b9271640b3b918b5cb65d9c68ea2c17d14319d2/csrc/fusion_segmenter.cpp#L355-L356. I was curious when the complete fusion has to be segmented (for acceptance by a scheduler) how nvFuser sets up I/O aliases of sub-fusions.

Reply:
I see.

The logic in segmenter is naive. Actually the whole codegen is pretty naive about I/O tensors. i.e. we pretend that we don't know about it at all and we just treat aliases as if they are new buffer.

The only thing the segmenter does is to add the alias-in as an input to the segment that's producing the alias-out. Just so that at runtime, we'll be able to use it for output.